flang/lib/Lower/IO.cpp

```diff
@@ -519,33 +550,39 @@ static void genInputItemList(Fortran::lower::AbstractConverter &converter,
     const auto *expr = Fortran::semantics::GetExpr(pVar);
     if (!expr)
       fir::emitFatalError(loc, "internal error: could not get evaluate::Expr");
-    auto itemTy = converter.genType(*expr);
-    auto inputFunc = getInputFunc(loc, builder, itemTy, isFormatted);
-    auto argType = inputFunc.getType().getInput(1);
-
-    llvm::SmallVector<mlir::Value> inputFuncArgs = {cookie};
-    if (argType.isa<fir::BoxType>()) {
-      if (Fortran::evaluate::HasVectorSubscript(*expr))
-        TODO(loc, "IO input item with vector subscripts");
-      auto box = fir::getBase(converter.genExprBox(*expr, stmtCtx, loc));
-      inputFuncArgs.push_back(builder.createConvert(loc, argType, box));
-    } else {
-      auto itemExv = converter.genExprAddr(expr, stmtCtx, loc);
-      auto itemAddr = fir::getBase(itemExv);
-      inputFuncArgs.push_back(builder.createConvert(loc, argType, itemAddr));
-      fir::factory::CharacterExprHelper charHelper{builder, loc};
-      if (charHelper.isCharacterScalar(itemTy)) {
-        auto len = fir::getLen(itemExv);
-        inputFuncArgs.push_back(
-            builder.createConvert(loc, inputFunc.getType().getInput(2), len));
-      } else if (itemTy.isa<mlir::IntegerType>()) {
-        inputFuncArgs.push_back(builder.create<mlir::ConstantOp>(
-            loc, builder.getI32IntegerAttr(
-                     itemTy.cast<mlir::IntegerType>().getWidth() / 8)));
+    if (Fortran::evaluate::HasVectorSubscript(*expr)) {
+      auto vectorSubscriptBox =
+          genVectorSubscriptBox(loc, converter, stmtCtx, *expr);
+      auto inputFunc = getInputFunc(
+          loc, builder, vectorSubscriptBox.getElementType(), isFormatted);
+      const bool mustBox = inputFunc.getType().getInput(1).isa<fir::BoxType>();
+      if (!checkResult) {
+        auto elementalGenerator = [&](const fir::ExtendedValue &element) {
+          createIoRuntimeCallForItem(loc, builder, inputFunc, cookie,
+                                     mustBox ? builder.createBox(loc, element)
+                                             : element);
+        };
+        vectorSubscriptBox.loopOverElements(builder, loc, elementalGenerator);
+      } else {
+        auto elementalGenerator =
+            [&](const fir::ExtendedValue &element) -> mlir::Value {
+          return createIoRuntimeCallForItem(
+              loc, builder, inputFunc, cookie,
+              mustBox ? builder.createBox(loc, element) : element);
+        };
+        if (!ok)
+          ok = builder.createBool(loc, true);
```

Comment:
I wonder if it is really worthwhile to lazily create the `true` constant... It doesn't hurt anything but it seems to invite bugs of omission.

Reply:
I think checkResult and ok could be merge into a single thing, the difference would be that an if ok/else would be generated for the first io-item, which is not the case today. I won't do it as part of this PR because it require changing more stuff unrelated to vector subscripts in IO (and will probably require some lit test updates). But if it sounds like to you something we should do, I can put a separate PR.

Reply:
Yes, that might work as well. There is no rush and bugs have higher priority.